Integrate riscv-compliance

sammer1107 opened the pull request:
+ This pull request integrate riscv-compliance (riscv-arch-test) 2.0 into rv32emu-next
+ The Makefile target support running all test cases or a specific test case. The command line options are explained in README.md
+ Some tests does not pass because it still doesn't have a correct value defined yet.

jserv commented:
Remove the generated files and rework!

jserv commented on CSRs.h:
```diff
@@ -0,0 +1,179 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
```
The range from Line 1 to 68 is not necessary. Remove these lines.

jserv replied:
?

kksweet8845 replied:
Sor, I may forget about it.

jserv commented:
Hint: use `git rebase -i` to rework these Git commits.

kksweet8845 commented:
Ok~

jserv commented:
> Sorry, I have a question.
> Do I need to rework all the commit messages?

Yes, all commits should be reworked. You shall squash the similar changes and make the commits clear.

kksweet8845 commented:
Hello, Jserv
I have reworked these commit messages.
Please check, thank you~

jserv commented:
After reworking Git commits with `git rebase -i`, you should perform `git push --force` for this branch.

jserv commented on Makefile:
```diff
@@ -1,19 +1,60 @@
 BINS = emu-rv32i test1
+TEST_TARGETS = \
```
Send another pull request for adding RISC-V test cases.

jserv commented:
Ensure no generated files such as ELF, log, dump files in Git repository.

jserv commented:
You have to rebase `master` branch of upstream. And then, force push.

jserv commented:
Don't use `git merge`. Use `git rebase` instead.

jserv commented:
Reference instructions:
1. set upstream: `git remote add upstream https://github.com/sysprog21/rv32emu && git fetch upstream`
2. Perform rebase: `git rebase upstream/master`
3. Fix the conflicts.

jserv commented on CSRs.h:
```diff
@@ -0,0 +1,179 @@
+
+#include <stdint.h>
+
+
```
Can you introduce a dedicated structure for preserving the following fields?

jserv commented on emu-rv32i.c:
```diff
@@ -1,3 +1,39 @@
+After this you can run it with the emulator like this:
+emu-rv32i zephyr/zephyr.elf
+original copyright:
+*/
```
Remove the range from Line 1 to Line 35.

jserv commented on emu-rv32i.c:
```diff
@@ -17,12 +53,19 @@
-// #define DEBUG_OUTPUT
-// #define DEBUG_EXTRA
+#define DEBUG_OUTPUT
+#define DEBUG_EXTRA
```
Don't do that.

jserv commented on emu-rv32i.c:
```diff
@@ -17,12 +53,19 @@
+#define DEBUG_EXTRA
+
+#if 1
+#define RV32C
```
`RV32C` should be assigned via Makefile

jserv commented on emu-rv32i.c:
```diff
@@ -33,87 +76,99 @@ uint32_t minmemr, maxmemr, minmemw, maxmemw;
-    ""};
+#define C_STATS_NUM 64
+
+#ifndef RV32C
```
I would like to use the following:
```c
char statnames[STATS_NUM + C_STATS_NUM][16] = {
    ... /* RV32I instructions */
#if defined(RV32C)
    ... /* RV32C specific instructions */
#endif
} 
```

kksweet8845 replied:
OK! It's more structural.

jserv commented on emu-rv32i.c:
```diff
@@ -33,87 +76,99 @@ uint32_t minmemr, maxmemr, minmemw, maxmemw;
-    maxmemr = maxmemw = 0;
+  for(int i=0;i<STATS_NUM;i++) stats[i]=0;
+#ifdef RV32C
+  for(int i=0;i<C_STATS_NUM;i++) stats[i+64] = 0;
```
Ditto. You can define macro for the range. Thus, we don't have to duplicate the loops.

jserv commented:
Before submitting the changes, ensure the completion of executing `clang-format -i *.[ch]`

jserv commented:
Reworked yet?

kksweet8845 commented:
Not yet,
I will update it tomorrow.

Jim Huang <notifications@github.com> 於 2020年1月17日 週五 上午1:10寫道：

> Reworked yet?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sysprog21/rv32emu/pull/5?email_source=notifications&email_token=AI4KYTPOEE7CU32GFLJLIVDQ6CIGPA5CNFSM4KFQ6NU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJE2HLQ#issuecomment-575251374>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI4KYTOCJ6566MQXVGGPNSDQ6CIGPANCNFSM4KFQ6NUQ>
> .
>

kksweet8845 commented:
Hello Jserv,
Sorry about the absence a few days ago.
I have rebased all commits.

jserv commented on .gitignore:
```diff
@@ -1,2 +1,9 @@
+sig.txt
+rom.v
+debug.txt
+CSR.h
```
`CSR.h` was tracked in GIT repository.

jserv commented on emu-rv32i.c:
```diff
@@ -23,6 +24,8 @@
 // #define DEBUG_EXTRA
 
+
+/* added by Shaos: */
```
You don't have to mention the contributor name inside source file.

jserv commented on emu-rv32i.c:
```diff
@@ -140,7 +163,7 @@ uint64_t mtimecmp;
 uint32_t ram_start;
 
-/* program entry point */
+/* Program entry point */
```
Don't do that. Let's minimize the changes.

jserv commented on emu-rv32i.c:
```diff
@@ -153,19 +176,26 @@ uint32_t end_signature = 0;
 uint32_t insn;
 uint32_t reg[32];
 
+
```
Ditto. Minimize the changes.

jserv commented:
Thank @kksweet8845 again for contributing to this project. However, we shall always merge commits with [better Git commit messages](https://chris.beams.io/posts/git-commit/) for further development. Please split this pull request into the following:
1. New pull request for rv32i conformance tests;
2. Once the above is merged, modify the pull request to concentrate on RV32C;

Changes regarding `CSR` instructions should be addressed in part 1.

sammer1107 committed:
Integrate riscv-compliance v2.0

+ Added --compliance option to rv32emu in order to dump
test signature.
+ Since the Makefile.include for different extensions
is the same, the folders in "test-rv32emu/device/rv32i_m" are all
symbolic links to "I".

sammer1107 committed:
Add README for running compliance

sammer1107 committed:
Update submodule to https protocol

Also rename test-rv32emu/work to build/compliance

jserv commented on Makefile:
```diff
@@ -60,6 +60,19 @@ check: $(BIN)
+# variables for compliance
+COMPLIANCE_DIR ?= ./riscv-compliance
+export RISCV_PREFIX ?= riscv32-unknown-elf-
+export RISCV_TARGET = test-rv32emu
```
Change the target to `compliance-tests` or something alike.

jserv commented on README.md:
```diff
@@ -30,6 +30,17 @@ make demo
 should appear when Doom is loaded and executed.
 
+### Run riscv-compliance test
+Once the submodule `riscv-compliance` is pulled, run all the available compliance test via command:
```
In `Makefile`, you can add submodule specific operations to pull automatically. See https://github.com/sysprog21/jitboy/blob/master/Makefile

jserv commented on test-rv32emu/device/rv32i_m/M:
```diff
@@ -0,0 +1 @@
+I
```
Why?

sammer1107 replied:
Because the file is the same as the file rv32i_m/I, so i used symbolic link.

jserv commented on test-rv32emu/device/rv32i_m/I/Makefile.include:
```diff
@@ -0,0 +1,23 @@
+    --compliance $(*).signature.output \
+    1>$(@) 2>&1
+
+RISCV_PREFIX   ?= riscv32-unknown-elf-
```
Why did you set `RISCV_PREFIX` here?

sammer1107 replied:
It's a mistake. I will fix that.

sammer1107 committed:
Modify makefile to pull submodule automatically

+ Also rename `test-rv32emu` to `compliance-target`
+ Remove unnecessary `RISCV_PREFIX` line in Makefile.include

jserv commented:
Can you avoid using symbolic links in source tree? In directory `build`, we can create symbolic links. However, if symbol links exists outside `build` directory, it is a bit difficult and confusing to remove the linked entries.

jserv commented on Makefile:
```diff
@@ -60,6 +60,22 @@ check: $(BIN)
+export TARGETDIR = $(shell pwd)
+export XLEN = 32
+export JOBS ?= -j
+export WORK = $(TARGETDIR)/build/compliance
```
I am not sure why `pwd` is necessary. Can you explain?

sammer1107 replied:
Because the build process of compliance involves multiple makefiles in different places, I found out that the makefile from the compliance will fail to find our target file if I just set `TARGETDIR` to `.`. Therefore, I set it to the absolute path of our project root.

jserv commented on main.c:
```diff
@@ -101,7 +105,8 @@ static void print_usage(const char *filename)
             "Options:\n"
-            "  --trace : print executable trace\n",
+            "  --trace : print executable trace\n"
+            "  --compliance [signature filename] : dump signature to the given file for compliance test\n",
```
Simplified as 
```
"  --compliance [filename] : dump signature to the given file, required bycompliance test\n",

```

sammer1107 committed:
Remove symbolic links in compliance target

+ Switch to using real directories and files.
+ Also simplified the help message for `--compliance` option

jserv commented:
Close in favor of #12
